app/code/Magento/CatalogSearch/Model/Adapter/Mysql/Aggregation/DataProvider/SelectBuilderForAttribute/StockConditionJoiner.php

```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\CatalogSearch\Model\Adapter\Mysql\Aggregation\DataProvider\SelectBuilderForAttribute;
+
+use Magento\CatalogInventory\Model\Stock;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\DB\Select;
+
+/**
+ * Join stock table with stock condition to select.
+ */
+class StockConditionJoiner
+{
+    /**
+     * @var ResourceConnection
+     */
+    private $resource;
+
+    /**
+     * @param ResourceConnection $resource
+     */
+    public function __construct(
+        ResourceConnection $resource
+    ) {
+        $this->resource = $resource;
+    }
+
+    /**
+     * @param Select $select
+     */
+    public function execute(Select $select)
+    {
+        $select->joinInner(
+            ['stock_index' => $this->resource->getTableName('cataloginventory_stock_status')],
+            'main_table.source_id = stock_index.product_id',
```

Comment:
please, test your code after writing at least manually. 
This condition looks very wrong to me:
```
‘main_table.source_id = stock_index.product_id’,
```